_posts/2020-09-06-how-the-system-tests-works.md

```diff
@@ -0,0 +1,596 @@
+            .addToConfig("offsets.topic.replication.factor", Math.min(kafkaReplicas, 3))
+            .addToConfig("transaction.state.log.min.isr", Math.min(kafkaReplicas, 2))
+            .addToConfig("transaction.state.log.replication.factor", Math.min(kafkaReplicas, 3))
+            .withNewListeners()
+                .withNewPlain().endPlain()
+                .withNewTls().endTls()
+            .endListeners()
+            .withNewInlineLogging()
+                .addToLoggers("log4j.rootLogger", "DEBUG")
+            .endInlineLogging()
+        .endKafka()
+        ...
+        ... (more configs...)
+        ... 
+    .endSpec()
+    .done()
+```
+
+In this case you can imagine that it is really easy to create multiple different instances of the `Kafka` instead of having separate objects.
+
+## Resources
+
+Resources in our way are abstraction to the `.yaml` definitions of custom resources which we provide such as 
+`Kafka`, `KafkaTopic`, `KafkaUser`, `KafkaConnect`, `KafkaMirrorMaker` and so on.
+These custom resources are encapsulated into java objects and some of you can remind the [ORM(Object Relation Mapping)](https://www.tutorialspoint.com/hibernate/orm_overview.htm) 
+from databases. 
+In the following Figure [2.1](#kafkatopic) we can see YAML representation of `KafkaTopic` and on the Code snippet 
+[2.2](kafkaimplementation) related java object.
+
+##### <a id="kafkatopic">Figure 2.1 YAML representation of `KafkaTopic`</a>
+```yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaTopic
+metadata:
+  name: my-topic
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  partitions: 1
+  replicas: 1
+  config:
+    retention.ms: 7200000
+    segment.bytes: 1073741824
+```
+
+##### <a id="kafkaimplementation">Code snippet 2.2 Loading the YAML file to JAVA representation object.</a>
+```java
+KafkaTopic kafkaTopic = getKafkaTopicFromYaml(PATH_TO_KAFKA_TOPIC_CONFIG); 
+// additional changes...
+kafkaTopic = new KafkaTopicBuilder(kafkaTopic)
+    .withNewMetadata()
+        .withName(topicName)
+        .withNamespace(ResourceManager.kubeClient().getNamespace())
+        .addToLabels(Labels.STRIMZI_CLUSTER_LABEL, clusterName)
+    .endMetadata()
+    .editSpec()
+        .withPartitions(partitions)
+        .withReplicas(replicas)
+        .addToConfig("min.insync.replicas", minIsr)
+    .endSpec();
+```
+
+Everything is build with three stacks. These three stacks are managed by `ResourceManager`, which is taking care of switching. 
+First one is `method stack` where his responsibility is to hold all resources, which are invoked in the test cases. 
+Second one is the `class stack` where his responsibility is to hold all resources, which are invoked in the `@BeforeAll` notation. 
+Switching between these two stacks are doing the third stack called pointer stack. 
+The pointer stack points to class stack where it is in the `@BeforeAll` notation otherwise we are working with the method stack. 
```

Comment:
Yes, we are always working with pointer stack. This sentencee is little bit complicated.

```
The `pointer stack` points to the `class stack` in we are in the `@BeforeAll` method, otherwise we are working with the `method stack`. 
```
Does this sentence make more sense to you? :)